### PR TITLE
Implement reduceRight function

### PR DESCRIPTION
**Pull request by DzyubSpirit**
Refs: #64

Copied from #83

### CODE REVIEW

**Comment by DzyubSpirit** [comment]
ping (polite)

File: test/array.reduceRight.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+const metasync = require('..');
+const metatests = require('metatests');
+
+metatests.test('successful with initial', test => {
```
Comment (SemenchenkoVitaliy) [inline test/array.reduceRight.js]:
```suggestion
metatests.test('reduceRight with initial', test => {
```

File: test/array.reduceRight.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    (prev, cur, callback) => process.nextTick(() => callback(null, prev + cur)),
+    (err, res) => {
+      test.strictSame(err, expectedError);
+      test.strictSame(res, undefined);
+      test.end();
+    }
+  );
+});
+
+metatests.test('successful without initial', test => {
```
Comment (SemenchenkoVitaliy) [inline test/array.reduceRight.js]:
```suggestion
metatests.test('reduceRight without initial', test => {
```

File: test/array.reduceRight.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    (prev, cur, callback) => process.nextTick(() => callback(null, prev + cur)),
+    (err, res) => {
+      test.error(err);
+      test.strictSame(res, expectedRes);
+      test.end();
+    }
+  );
+});
+
+metatests.test('successful with asymetric function', test => {
```
Comment (SemenchenkoVitaliy) [inline test/array.reduceRight.js]:
```suggestion
metatests.test('reduceRight with asymetric function', test => {
```

File: test/array.reduceRight.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      process.nextTick(() => callback(null, prev * 2 + +cur)),
+    (err, res) => {
+      test.error(err);
+      test.strictSame(res, expectedRes);
+      test.end();
+    }
+  );
+});
+
+metatests.test('with error', test => {
```
Comment (SemenchenkoVitaliy) [inline test/array.reduceRight.js]:
```suggestion
metatests.test('reduceRight with error', test => {
```

File: lib/array.js [hunk 5]
```diff
@@ -201,6 +201,53 @@ const reduce = (items, fn, done, initial) => {
...
+//     result - <Array>
+//   initial - <any>, optional value to be used as first
+//       argument in first iteration
+const reduceRight = (items, fn, done, initial) => {
+  done = done || common.emptyness;
+  const len = items.length;
+  let count = typeof initial === 'undefined' ? len - 2 : len - 1;
+
+  if (!len) {
+    const err = count === len - 2 ? new TypeError(EMPTY_ARR) : null;
```
Comment (SemenchenkoVitaliy) [inline lib/array.js]: I think we need a different error message for `reduceRight`: `'Metasync: reduceRight of the empty array with no initial value'` though I don't have strong opinion about this.

File: lib/array.js [hunk 6]
```diff
@@ -201,6 +201,53 @@ const reduce = (items, fn, done, initial) => {
   fn(previous, current, next, count, items);
 };
 
+// Asynchronous reduceRight
+//   items - <Array>, incoming
+//   fn - <Function>, to be executed for each value in array
+//     previous - <any>, value previously returned in the last iteration
+//     current - <any>, current element being processed in the array
+//     callback - <Function>, callback for returning value
+//         back to reduce function
```
Comment (SemenchenkoVitaliy) [inline lib/array.js]:
```suggestion
//         back to reduce function
//       err - <Error> | <null>
//       data - <any>, resulting value
```

File: lib/array.js [hunk 7]
```diff
@@ -201,6 +204,59 @@ const reduce = (items, fn, done, initial) => {
...
+//     items - <Array>, the array reduce was called upon
+//   done - <Function>, on done, optional
+//     err - <Error> | <null>
+//     result - <Array>
+//   initial - <any>, optional value to be used as first
+//       argument in first iteration
+const reduceRight = (items, fn, done, initial) => {
+  done = done || common.emptyness;
+  const len = items.length;
+  let count = typeof initial === 'undefined' ? len - 2 : len - 1;
```
Comment (nechaido) [inline lib/array.js]: What about array with one element and no initial value? `count` would equal `-1`.

File: lib/array.js [hunk 8]
```diff
@@ -151,7 +151,8 @@ const filter = (items, fn, done) => {
   }
 };
 
-const EMPTY_ARR = 'Metasync: reduce of empty array with no initial value';
+const REDUCE_EMPTY_ARR =
```
Comment (nechaido) [inline lib/array.js]: This and change and the change in the doc comment are semi-relevant, I am fine with having them in this PR, though I would like to extract them into a separate commit.

File: lib/array.js (hunk 7 above)
Comment (DzyubSpirit) [reply lib/array.js]: What a sharp eye! Many thanks for that :100: 
I will fix it here, fix reduce and add more tests for both functions

File: lib/array.js (hunk 8 above)
Comment (DzyubSpirit) [reply lib/array.js]: I will leave 3 commits then:
1. reduceRight and tests for it.
2. Fix of the reduce.
3. Change of the comments and errors.

File: lib/array.js (hunk 7 above)
Comment (DzyubSpirit) [reply lib/array.js]: Done

File: lib/array.js (hunk 8 above)
Comment (DzyubSpirit) [reply lib/array.js]: Done

**Comment by DzyubSpirit** [comment]
I've done squashing redundant commits so I think the branch is ready for merge.
I'm not authorized to merge so @nechaido, @tshemsedinov or @belochub, please, do it for me.

File: lib/array.js [hunk 9]
```diff
@@ -171,15 +174,20 @@ const EMPTY_ARR = 'Metasync: reduce of empty array with no initial value';
...
   }
 
-  let previous = count === 1 ? items[0] : initial;
+  let previous = hasInitial ? initial : items[0];
+  if ((len === 0 && hasInitial) || (len === 1 && !hasInitial)) {
+    done(null, previous);
+    return;
+  }
+
+  let count = 1 - hasInitial;
```
Comment (nechaido) [inline lib/array.js]: I am strongly against implicit boolean to number conversion.

File: lib/array.js [hunk 10]
```diff
@@ -201,6 +209,64 @@ const reduce = (items, fn, done, initial) => {
...
+    return;
+  }
+
+  let previous = hasInitial ? initial : items[len - 1];
+  if ((len === 0 && hasInitial) || (len === 1 && !hasInitial)) {
+    done(null, previous);
+    return;
+  }
+
+  let count = len - 2 + hasInitial;
```
Comment (nechaido) [inline lib/array.js]: Ditto.

File: lib/array.js (hunk 9 above)
Comment (DzyubSpirit) [reply lib/array.js]: Done

File: lib/array.js (hunk 10 above)
Comment (DzyubSpirit) [reply lib/array.js]: Ditto.

**Comment by DzyubSpirit** [comment]
@nechaido, that definitely looks better though it's a bit of hassle requiring not just renaming but also moving some changes between commits.
I'll do it
